lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+)
+
+// event represents detected changes at one path until it times out and a scan is scheduled
+type event struct {
+	path         string
+	firstModTime time.Time
+	lastModTime  time.Time
+	evType       eventType
+}
+
+type eventBatch map[string]*event
+
+// Stores both events directly within this directory (events) and child
+// directories recursively containing events themselves (dirs). The keys to the
+// events and dirs maps are the full relative path to the Syncthing folder root.
+type eventDir struct {
+	path      string
+	parentDir *eventDir
+	events    eventBatch
+	dirs      map[string]*eventDir
+}
+
+func newEventDir(path string, parentDir *eventDir) *eventDir {
+	return &eventDir{
+		path:      path,
+		parentDir: parentDir,
+		events:    make(eventBatch),
+		dirs:      make(map[string]*eventDir),
+	}
+}
+
+type watcher struct {
+	folderID            string
+	folderPath          string
+	folderIgnores       *ignore.Matcher
+	folderIgnoresUpdate chan *ignore.Matcher
+	folderCfg           config.FolderConfiguration
+	// Time after which an event is scheduled for scanning when no modifications occur.
+	notifyDelay time.Duration
+	// Time after which an event is scheduled for scanning even though modifications occur.
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	notifyTimerResetChan  chan time.Duration
+	notifyChan            chan []string
+	// All detected and to be scanned events are stored in a tree like
+	// structure mimicking directories to keep count of events per directory.
+	rootEventDir     *eventDir
+	backendEventChan chan notify.EventInfo
+	// paths currently processed by Syncthing (i.e. ignore modifications on them)
+	inProgress   map[string]struct{}
+	cfg          *config.Wrapper
+	configUpdate chan config.FolderConfiguration
+	stop         chan struct{}
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	C() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+	VerifyConfiguration(from, to config.Configuration) error
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(id string, cfg *config.Wrapper, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &watcher{
+		folderID:              id,
+		folderIgnores:         ignores,
+		folderIgnoresUpdate:   make(chan *ignore.Matcher),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		backendEventChan:      make(chan notify.EventInfo, maxFiles),
+		inProgress:            make(map[string]struct{}),
+		cfg:                   cfg,
+		configUpdate:          make(chan config.FolderConfiguration),
+		stop:                  make(chan struct{}),
+	}
+	folderCfg, ok := cfg.Folder(id)
+	if !ok {
+		panic(fmt.Sprintf("bug: Folder %s does not exist", id))
+	}
+	fsWatcher.updateConfig(folderCfg)
+
+	if err := fsWatcher.setupBackend(); err != nil {
```

Comment:
Conceptually, this should be done run in Serve(), as we start doing work early, even if the service is not started.
Also, it would allow to keep the execution loop tight (in one function) and might allow us to remove some members from the struct, as for example backendEventChan might only need to live within Serve() and passed to child functions, etc, etc.